<launch>
  <!--
  ===================================================================================
  |      Main Bringup Launch File for the Bessica-D Robot with MoveIt!              |
  ===================================================================================
  -->
  <arg name="moveit_controller_manager" default="moveit_simple_controller_manager/MoveItSimpleControllerManager"/>
  <!-- Load the URDF, SRDF and other .yaml configuration files on the param server -->
  <include file="$(find bessica_d_moveit)/launch/planning_context.launch">
    <arg name="load_robot_description" value="true"/>
  </include>

  <!-- Load the ROS controllers configuration -->
  <rosparam file="$(find bessica_d_driver)/config/controllers.yaml" command="load" />
  
  <!-- Load the MoveIt controller connection file -->
  <rosparam file="$(find bessica_d_moveit)/config/simple_moveit_controllers.yaml"/>

  <!-- =================================================================================== -->
  <!-- |                       START THE HARDWARE AND ROS_CONTROL                          | -->
  <!-- =================================================================================== -->

  <!-- Start the low-level serial driver node -->
  <node name="bessica_d_driver_node" pkg="bessica_d_driver" type="bessica_d_driver_node" output="screen">
    <param name="port"           type="string" value="/dev/ttyCH341USB0"/>
    <param name="baud_rate"      type="int"    value="921600"/>
    <param name="debug_mode"     type="bool"   value="false"/>
  </node>
  
  <!-- Load the joint names for the hardware interface -->
  <rosparam file="$(find bessica_d_driver)/config/joint_names.yaml" command="load" />
  
  <!-- Start the ros_control hardware interface node -->
  <node name="bessica_d_hw_interface" pkg="bessica_d_driver" type="bessica_d_hw_interface_node" output="screen"/>

  <!-- Load the robot state publisher. No remapping needed now. -->
  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" respawn="true" output="screen" />
  
  <!-- Load the controller spawner in the global namespace -->
  <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false"
    output="screen" args="
      joint_state_controller
      left_arm_controller
      right_arm_controller
      left_hand_controller
      right_hand_controller
    "/>

  <!-- =================================================================================== -->
  <!-- |                                  START MOVEIT!                                  | -->
  <!-- =================================================================================== -->
  
  <!-- Run the main MoveIt! planning execution context -->
  <include file="$(find bessica_d_moveit)/launch/move_group.launch">
    <arg name="load_robot_description" value="false"/>
    <!-- This param tells move_group to load the ros_control plugin -->
    <param name="moveit_controller_manager" value="$(arg moveit_controller_manager)" />
  </include>

  <!-- =================================================================================== -->
  <!-- |                                 START VISUALIZATION                             | -->
  <!-- =================================================================================== -->

  <!-- Run RViz and load the default config to see the state of the move_group node -->
  <include file="$(find bessica_d_moveit)/launch/moveit_rviz.launch">
    <arg name="rviz_config" value="$(find bessica_d_moveit)/launch/moveit.rviz"/>
  </include>

</launch>